sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/implementation/OkHttpToAzureCoreHttpHeadersWrapper.java

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * Wraps an OkHttp HttpHeaders instance and provides an azure-core HttpHeaders view onto it. This avoids the need to
+ * copy the OkHttp HttpHeaders into an azure-core HttpHeaders instance. Whilst it's not necessary to support mutability
+ * (as these headers are the result of an OkHttp response), we do so in any case, given the additional implementation
+ * cost is minimal.
+ */
+public final class OkHttpToAzureCoreHttpHeadersWrapper extends HttpHeaders {
+    private final Headers okhttpHeaders;
+
+    private HttpHeaders azureCoreHeaders;
+    private boolean converted = false;
+
+    public OkHttpToAzureCoreHttpHeadersWrapper(Headers okhttpHeaders) {
+        this.okhttpHeaders = okhttpHeaders;
+        this.azureCoreHeaders = new HttpHeaders(okhttpHeaders.size() * 2);
+    }
+
+    @Override
+    public int getSize() {
+        return converted ? azureCoreHeaders.getSize() : okhttpHeaders.size();
+    }
+
+    @Override
+    public HttpHeaders add(String name, String value) {
+        if (name == null || value == null) {
+            return this;
+        }
+
+        convertIfNeeded();
+
+        azureCoreHeaders.add(name, value);
+        return this;
+    }
+
+    @Override
+    public HttpHeaders set(String name, String value) {
+        if (name == null) {
+            return this;
+        }
+
+        convertIfNeeded();
+
+        azureCoreHeaders.set(name, value);
+        return this;
+    }
+
+    @Override
+    public HttpHeaders set(String name, List<String> values) {
+        if (name == null) {
+            return this;
+        }
+
+        convertIfNeeded();
+
+        azureCoreHeaders.set(name, values);
+        return this;
+    }
+
+    @Override
+    public HttpHeaders setAll(Map<String, List<String>> headers) {
+        convertIfNeeded();
+
+        azureCoreHeaders.setAll(headers);
+        return this;
+    }
+
+    @Override
+    public HttpHeader get(String name) {
+        convertIfNeeded();
+
+        return azureCoreHeaders.get(name);
+    }
+
+    @Override
+    public HttpHeader remove(String name) {
+        convertIfNeeded();
+
+        return azureCoreHeaders.remove(name);
+    }
+
+    @Override
+    public String getValue(String name) {
+        convertIfNeeded();
+
+        return azureCoreHeaders.getValue(name);
+    }
+
+    @Override
+    public String[] getValues(String name) {
+        convertIfNeeded();
+
+        return azureCoreHeaders.getValues(name);
+    }
+
+    @Override
+    public Map<String, String> toMap() {
+        convertIfNeeded();
+
+        return azureCoreHeaders.toMap();
+    }
+
+    @Override
+    public Iterator<HttpHeader> iterator() {
+        convertIfNeeded();
+
+        return azureCoreHeaders.iterator();
+    }
+
+    @Override
+    public Stream<HttpHeader> stream() {
+        convertIfNeeded();
+
+        return azureCoreHeaders.stream();
+    }
+
+    @Override
+    public String toString() {
+        convertIfNeeded();
```

Comment:
is possible we only call convertIfNeeded() once in constructor?

Reply:
Calling `convertIfNeeded` in the constructor would defeat the purpose of deferring the conversion of OkHttp's `Headers` to `azure-core`'s `HttpHeaders` as it would be converted eagerly.